Python/symtable.c

```diff
@@ -516,21 +544,34 @@ analyze_name(PySTEntryObject *ste, PyObject *scopes, PyObject *name, long flags,
                          "nonlocal declaration not allowed at module level");
             return error_at_directive(ste, name);
         }
-        if (!PySet_Contains(bound, name)) {
+        PyObject *flags = PyDict_GetItem(bound, name);
+        if (flags == NULL) {
             PyErr_Format(PyExc_SyntaxError,
                          "no binding for nonlocal '%U' found",
                          name);
 
             return error_at_directive(ste, name);
         }
+        if (PyLong_AsLong(flags) & DEF_TYPE_PARAM) {
+            PyErr_Format(PyExc_SyntaxError,
+                         "nonlocal binding not allowed for type parameter '%U'",
+                         name);
+            return error_at_directive(ste, name);
+        }
         SET_SCOPE(scopes, name, FREE);
         ste->ste_free = 1;
         return PySet_Add(free, name) >= 0;
     }
     if (flags & DEF_BOUND) {
         SET_SCOPE(scopes, name, LOCAL);
-        if (PySet_Add(local, name) < 0)
+        PyObject *flags_obj = PyLong_FromLong(flags);
+        if (flags_obj == NULL)
+            return 0;
+        if (PyDict_SetItem(local, name, flags_obj) < 0) {
```

Comment:
Yes, I wasn't happy with this change, especially because it's only needed for a fairly marginal check (we want to disallow use of `nonlocal` to modify type params). I can try out your approach, but a possible difficulty is that we want to allow cases where there is an inner, non-type param variable shadowing the type param:

```
def f[T, U]():
    U = int
    def inner():
        nonlocal T # should not be allowed
        nonlocal U # should be allowed
```

Probably that's doable by discarding items from the typeparams set that are bound in the middle scope.

I also realized this check is buggy: there is a test `test_name_non_collision_11` in test_type_params that should be an error but isn't. I'll have to get back to this later.

Reply:
Got your suggestion to work, and now it correctly errors on `test_name_non_collision_11`.

Relatedly, I got worried that in the symtable I had the same issue as you pointed out in the compiler, where I didn't exit scopes on various error paths. But apparently that's not necessary in the symtable (compare existing code at https://github.com/python/cpython/blob/29f348e232e82938ba2165843c448c2b291504c5/Python/symtable.c#L1295), because exiting a scope there doesn't directly call free(), it just removes an item from a list, and presumably if there is an error, eventually the whole list gets collected and everything is freed as expected. Summary: I think we're good.